animeal/src/Flows/Main/Modules/Favourites/Favourites/ViewModel/FavouritesViewModel.swift

```diff
@@ -0,0 +1,100 @@
+import UIComponents
+
+final class FavouritesViewModel: FavouritesViewModelLifeCycle, FavouritesViewInteraction, FavouritesViewState {
+    // MARK: - Private properties
+    private lazy var shimmerScheduler = ShimmerViewScheduler()
+    
+    // MARK: - Dependencies
+    private let model: FavouritesModelProtocol
+    private let coordinator: FavouritesCoordinatable
+    private let mapper: FavouriteViewItemMappable
+
+    // MARK: - State
+    var onErrorIsNeededToDisplay: ((String) -> Void)?
+    var onContentHaveBeenPrepared: ((FavouriteViewContentState) -> Void)?
+    var onMediaContentHaveBeenPrepared: ((FavouriteMediaContent) -> Void)?
+
+    // MARK: - Initialization
+    init(
+        coordinator: FavouritesCoordinatable,
+        mapper: FavouriteViewItemMappable = FavouriteViewItemMapper(),
+        model: FavouritesModelProtocol
+    ) {
+        self.coordinator = coordinator
+        self.mapper = mapper
+        self.model = model
+    }
+    
+    // MARK: - Life cycle
+    func load() {
+        shimmerScheduler.start()
+        updateViewItems { [weak self] in
+            self?.updateViewLoadingItems() ?? []
+        }
+        updateViewItems { [weak self] in
+            guard let self else { return [] }
+            self.shimmerScheduler.stop()
+            return try await self.updateViewContentItems()
+        }
+    }
+    
+    private func loadMediaContent(_ feedingPointId: String, _ key: String?) {
+        guard let key = key else { return }
+        model.fetchMediaContent(key: key) { [weak self] content in
+            guard let self = self else { return }
+            if let mediaContent = self.mapper.mapFeedingPointMediaContent(feedingPointId, content) {
+                self.onMediaContentHaveBeenPrepared?(mediaContent)
+            }
+        }
+    }
+    
+    private func updateViewLoadingItems() -> [FavouriteItem] {
+        let loadingItems = (0...1).map { _ in
+            FavouriteShimmerViewItem(scheduler: shimmerScheduler)
+        }
+        
+        return loadingItems.map { mapper.mapShimmerViewItem($0) }
+    }
+    
+    private func updateViewContentItems() async throws -> [FavouriteItem] {
+        let favourites = try await model.fetchFavourites()
+        favourites.forEach { content in
+            self.loadMediaContent(content.feedingPointId, content.header.cover)
+        }
+        return favourites.map { self.mapper.mapFavourite($0) }
+    }
+    
+    private func updateViewItems(
+        _ operation: @escaping () async throws -> [FavouriteItem]
+    ) {
+        Task { [weak self] in
+            do {
+                let viewItems = try await operation()
+
+                guard !viewItems.isEmpty else {
+                    DispatchQueue.main.async { [weak self] in
+                        self?.onContentHaveBeenPrepared?(.empty(L10n.Favourites.empty))
+                    }
+                    return
+                }
+
+                DispatchQueue.main.async { [weak self] in
```

Comment:
If we use async/await approach, it will be better to use @MainActor instead of GCD

Reply:
You could use main actor for the whole scope, method or class

Reply:
I marked the protocol as main actor to be sure that async based methods will be return to main

Reply:
Hello @diantyn thanks you're right; i'm going to create another PR with these fixes